### ISSUE TITLE
BUG: Speed degradation with categoricals

### ISSUE THREAD

**Issue by RedTailedHawk** [pandas-dev/pandas#35546]
I have noticed some major speed degradation when using categorical data types.  For example, a data frame with 1 million rows that sums 10 float columns and groups by two columns (one a date column and one a category column).  The cardinality of the category seems to have a major effect.  When grouping on category column of cardinality 10, performance is decent (query runs in 150 ms).  But with cardinality of 100, the query runs in 10 seconds.  If I re-generate the data frame without  categorical columns, the same query that took 10 seconds with categoricals now runs in 350 ms.

**Comment by RedTailedHawk** [comment]
```
import datetime as dt
import random

import numpy as np
import pandas as pd
import pyarrow as pa

pd.options.display.float_format = '{:,.4f}'.format

print("Pandas v" + pd.__version__)
print("PyArrow v" + pa.__version__)

use_categoricals = True
dim_col = "a2"

print("\nGenerating data with{} categoricals...".format("out" if use_categoricals else ""))

NUM_ROWS = int(1e6)

def random_datetimes_or_dates(start, end, out_format='date', n=10):
    (divide_by, unit) = (10**9, 's') if out_format=='datetime' else (24*60*60*10**9, 'D')
    start_u = start.value//divide_by
    end_u = end.value//divide_by

    return pd.to_datetime(np.random.randint(start_u, end_u, n), unit=unit)

start = pd.to_datetime('2015-01-01')
end = pd.to_datetime('2018-01-01')

NUM_ATTRIBS = 9    # plus 1 for date
NUM_METRICS = 20
ATTR_CHAR_LEN = 20
VAL_LOW = 0
VAL_HIGH = 100

dim1 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(10, 1))
dim2 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(100, 1))
dim3 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(100, 1))
dim4 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(1000, 1))
dim5 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(1000, 1))
dim6 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(10000, 1))
dim7 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(10000, 1))
dim8 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(100000, 1))
dim9 = pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(1000000, 1))

attr_col_names = [f'a{i+1}' for i in range(NUM_ATTRIBS)]
metric_col_names = [f'm{i+1}' for i in range(NUM_METRICS)]

df_attr = pd.DataFrame(pd.util.testing.rands_array(nchars=ATTR_CHAR_LEN, size=(NUM_ROWS, NUM_ATTRIBS)), columns=attr_col_names)
df_attr.insert(loc=0, column='date', value=random_datetimes_or_dates(start, end, out_format='date', n=NUM_ROWS))

df_attr["a1"] = df_attr["a1"].apply(lambda x: dim1[random.randint(0, len(dim1) - 1)][0])
df_attr["a2"] = df_attr["a2"].apply(lambda x: dim2[random.randint(0, len(dim2) - 1)][0])
df_attr["a3"] = df_attr["a3"].apply(lambda x: dim3[random.randint(0, len(dim3) - 1)][0])
df_attr["a4"] = df_attr["a4"].apply(lambda x: dim4[random.randint(0, len(dim4) - 1)][0])
df_attr["a5"] = df_attr["a5"].apply(lambda x: dim5[random.randint(0, len(dim5) - 1)][0])
df_attr["a6"] = df_attr["a6"].apply(lambda x: dim6[random.randint(0, len(dim6) - 1)][0])
df_attr["a7"] = df_attr["a7"].apply(lambda x: dim7[random.randint(0, len(dim7) - 1)][0])
df_attr["a8"] = df_attr["a8"].apply(lambda x: dim8[random.randint(0, len(dim8) - 1)][0])
df_attr["a9"] = df_attr["a9"].apply(lambda x: dim9[random.randint(0, len(dim9) - 1)][0])

df_metrics = pd.DataFrame(np.random.randint(VAL_LOW, VAL_HIGH, size=(NUM_ROWS, NUM_METRICS)), columns=metric_col_names).astype(float)

df = pd.concat([df_attr, df_metrics], axis=1, sort=False)

if use_categoricals:
    df["a1"] = df["a1"].astype("category")
    df["a2"] = df["a2"].astype("category")
    df["a3"] = df["a3"].astype("category")
    df["a4"] = df["a4"].astype("category")
    df["a5"] = df["a5"].astype("category")
    df["a6"] = df["a6"].astype("category")
    df["a7"] = df["a7"].astype("category")
    df["a8"] = df["a8"].astype("category")
    df["a9"] = df["a9"].astype("category")

print("\nRow count = {}".format(len(df)))
print("Dim column cardinality = {}".format(df[dim_col].nunique()))

ts0 = dt.datetime.now()

df = df.groupby(["date", dim_col]).agg(m1=("m1", "sum"), m2=("m2", "sum"), m3=("m3", "sum"), m4=("m4", "sum"), m5=("m5", "sum"), m6=("m6", "sum"), m7=("m7", "sum"), m8=("m8", "sum"), m9=("m9", "sum"), m10=("m10", "sum"))

ts1 = dt.datetime.now()
print("\nQuery time (ms) = " + str(int((ts1 - ts0).total_seconds() * 1000)))

print("\nResult set row count = {}".format(len(df)))
print(df.head(10))
```

**Comment by RedTailedHawk** [comment]
![image](https://user-images.githubusercontent.com/8505928/89310023-2ed33280-d642-11ea-93e3-cfd1ea198f28.png)

**Comment by RedTailedHawk** [comment]
Note that the above screenshot is from my post on the PyArrow JIRA forum, but the Pandas code above takes Parquet out of the loop.

For reference:  https://issues.apache.org/jira/browse/ARROW-9637

**Comment by RedTailedHawk** [comment]
Also, I upgraded to Pandas 1.1.0, and the query time is even slower.

Thanks for any help.
-Larry

**Comment by jorisvandenbossche** [comment]
@RedTailedHawk thanks for the report!

A follow-up question: can you try to simplify a bit the reproducible example? Right now it is hard to follow because of the many variables (and takes a long time to create), while not all columns are used for the actual line that is timed. I think it would also simplify if you use a single "a" categorical column (that can be re-created with a different cardinality if you want to time that), instead of creating 9 of them (since only 1 is used)

**Comment by RedTailedHawk** [comment]
@jorisvandenbossche Sure, I just updated the code example above to only use two categorical columns.  Should speed things up for you.

**Comment by jorisvandenbossche** [comment]
Further simplified example:

```python
import numpy as np
import pandas as pd

NUM_ROWS = int(1e6)
CARDINALITY = 10

start = pd.to_datetime('2015-01-01')
end = pd.to_datetime('2018-01-01')
categories = pd.util.testing.rands_array(nchars=20, size=CARDINALITY)

df = pd.DataFrame({
    "date": pd.to_datetime(np.random.randint(start.value, end.value, NUM_ROWS)).round("D"),
    "cat": categories[np.random.randint(0, len(categories), size=NUM_ROWS)],
    "m": np.random.randint(0, 100, size=NUM_ROWS).astype(float),
})

# no categorical dtype
%timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))

# with categorical dtype
df_cat = df.copy()
df_cat["cat"] = df_cat["cat"].astype("category")

%timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))
```

On pandas 1.0.5

```
In [3]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))  
138 ms ± 11.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum")) 
1.57 s ± 318 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

With master (basically 1.1.0):

```
In [2]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))       
161 ms ± 28.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))   
56 ms ± 3.54 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Increasing the cardinality to 100, with pandas 1.0.5:

```
In [7]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum")) 
156 ms ± 40.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))
754 ms ± 132 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

and with master / 1.1.0
```
In [10]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))  
170 ms ± 14.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [12]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))                                                                                                                                               
2.85 s ± 335 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So with small cardinaity, it seems to have improved quite a bit, but for large cardinality the performance degraded in 1.1.

I know there have been previous issues about groupby performance issues with categorical columns, but can't directly remember which ones have been tackled in 1.1.

**Comment by RedTailedHawk** [comment]
OK, thank you for looking at this!

**Comment by simonjayhawkins** [comment]
moved off 1.1.1 milestone (scheduled for this week) as no PRs to fix in the pipeline

**Comment by simonjayhawkins** [comment]
moved off 1.1.2 milestone (scheduled for this week) as no PRs to fix in the pipeline

**Comment by simonjayhawkins** [comment]
moved off 1.1.3 milestone (overdue) as no PRs to fix in the pipeline

**Comment by simonjayhawkins** [comment]
moved off 1.1.4 milestone (scheduled for release tomorrow) as no PRs to fix in the pipeline

**Comment by NowanIlfideme** [comment]
Also noticed this performance issue. Given that CategoricalDType is backed by integers, this is pretty surprising behavior... even converting the values from category to string brings the time of a groupby operation from 20 seconds to <0.1s...

**Comment by jreback** [comment]
u need to add observed=True

**Comment by jbrockmendel** [comment]
Using joris's reproducer https://github.com/pandas-dev/pandas/issues/35546#issuecomment-668791892 with CARDINALITY=100 I'm getting

```
In [14]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))
133 ms ± 5.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [15]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))
163 ms ± 1.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

so this may be resolved.  Could use a double-check.

**Comment by mroeschke** [comment]
On main, categorical appears faster now so I think this has been addressed

```
In [10]: %timeit df_cat.groupby(["date", "cat"]).agg(m=("m", "sum"))
46.3 ms ± 302 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [11]: %timeit df.groupby(["date", "cat"]).agg(m=("m", "sum"))
89 ms ± 3.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
